clang/test/CodeGenSYCL/launch_bounds_nvptx.cpp

```diff
@@ -0,0 +1,105 @@
+
+#include "sycl.hpp"
+
+using namespace sycl;
+queue q;
+
+class Foo {
+public:
+  [[intel::max_work_group_size(8, 8, 8), intel::min_work_groups_per_cu(2),
+    intel::max_work_groups_per_mp(4)]] void
+  operator()() const {}
+};
+
+template <int N> class Functor {
+public:
+  [[intel::max_work_group_size(N, 8, 8), intel::min_work_groups_per_cu(N),
+    intel::max_work_groups_per_mp(N)]] void
+  operator()() const {}
+};
+
+template <int N>
+[[intel::max_work_group_size(N, 8, 8), intel::min_work_groups_per_cu(N),
+  intel::max_work_groups_per_mp(N)]] void
+zoo() {}
+
+[[intel::max_work_group_size(8, 8, 8), intel::min_work_groups_per_cu(2),
+  intel::max_work_groups_per_mp(4)]] void
+bar() {}
+
+int main() {
+  q.submit([&](handler &h) {
+    // Test attribute argument size.
+    Foo boo;
+    h.single_task<class kernel_name1>(boo);
+
+    // Test attribute is applied on lambda.
+    h.single_task<class kernel_name2>(
+        [] [[intel::max_work_group_size(8, 8, 8),
+             intel::min_work_groups_per_cu(2),
+             intel::max_work_groups_per_mp(4)]] () {});
+
+    // Test class template argument.
+    Functor<6> f;
+    h.single_task<class kernel_name3>(f);
+
+    // Test attribute is propagated.
+    h.single_task<class kernel_name4>([]() { bar(); });
+
+    // Test function template argument.
+    h.single_task<class kernel_name5>([]() { zoo<16>(); });
+  });
+  return 0;
+}
+
+// CHECK: define dso_local void @{{.*}}kernel_name1() #0 {{.*}} !min_work_groups_per_cu ![[MWGPC:[0-9]+]] !max_work_groups_per_mp ![[MWGPM:[0-9]+]] !max_work_group_size ![[MWGS:[0-9]+]]
+// CHECK: define dso_local void @{{.*}}kernel_name2() #0 {{.*}} !min_work_groups_per_cu ![[MWGPC:[0-9]+]] !max_work_groups_per_mp ![[MWGPM:[0-9]+]] !max_work_group_size ![[MWGS:[0-9]+]]
+// CHECK: define dso_local void @{{.*}}kernel_name3() #0 {{.*}} !min_work_groups_per_cu ![[MWGPC_MWGPM:[0-9]+]] !max_work_groups_per_mp ![[MWGPC_MWGPM]] !max_work_group_size ![[MWGS_2:[0-9]+]]
+// CHECK: define dso_local void @{{.*}}kernel_name4() #0 {{.*}} !min_work_groups_per_cu ![[MWGPC:[0-9]+]] !max_work_groups_per_mp ![[MWGPM:[0-9]+]] !max_work_group_size ![[MWGS:[0-9]+]]
+// CHECK: define dso_local void @{{.*}}kernel_name5() #0 {{.*}} !min_work_groups_per_cu ![[MWGPC_MWGPM_2:[0-9]+]] !max_work_groups_per_mp ![[MWGPC_MWGPM_2]] !max_work_group_size ![[MWGS_3:[0-9]+]]
+
+// CHECK: {{.*}}@{{.*}}kernel_name1, !"maxntidx", i32 512}
+// CHECK: {{.*}}@{{.*}}kernel_name1, !"minnctapersm", i32 2}
+// CHECK: {{.*}}@{{.*}}kernel_name1, !"maxclusterrank", i32 4}
+// CHECK: {{.*}}@{{.*}}Foo{{.*}}, !"maxntidx", i32 512}
+// CHECK: {{.*}}@{{.*}}Foo{{.*}}, !"minnctapersm", i32 2}
+// CHECK: {{.*}}@{{.*}}Foo{{.*}}, !"maxclusterrank", i32 4}
+// CHECK: {{.*}}@{{.*}}kernel_name2, !"maxntidx", i32 512}
+// CHECK: {{.*}}@{{.*}}kernel_name2, !"minnctapersm", i32 2}
+// CHECK: {{.*}}@{{.*}}kernel_name2, !"maxclusterrank", i32 4}
+// CHECK: {{.*}}@{{.*}}main{{.*}}, !"maxntidx", i32 512}
```

Comment:
What does this correspond to? The lambda?

Reply:
`_ZZZ4mainENKUlRN4sycl3_V17handlerEE_clES2_ENKUlvE_clEv`